sound/soc/sof/ipc4.c

```diff
@@ -0,0 +1,244 @@
+					u32 extension, void *msg_data, size_t msg_bytes,
+					void *reply_data, size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	if (ipc->disable_ipc_tx)
+		return -ENODEV;
+
+	/*
+	 * The spin-lock is also still needed to protect message objects against
+	 * other atomic contexts.
+	 */
+	spin_lock_irq(&sdev->ipc_lock);
+
+	/* initialise the message */
+	msg = &ipc->msg;
+
+	msg->header = header;
+	msg->extension = extension;
+	msg->msg_size = msg_bytes;
+	msg->reply_size = reply_bytes;
+	msg->reply_error = 0;
+
+	/* attach any data */
+	if (msg_bytes > msg->max_payload_size || reply_bytes > msg->max_payload_size)
+		return -EINVAL;
+
+	memcpy(msg->msg_data, msg_data, msg_bytes);
+
+	sdev->msg = msg;
+
+	ret = snd_sof_dsp_send_msg(sdev, msg);
+	/* Next reply that we receive will be related to this message */
+	if (!ret)
+		msg->ipc_complete = false;
+
+	spin_unlock_irq(&sdev->ipc_lock);
+
+	if (ret) {
+		dev_err_ratelimited(sdev->dev, "ipc tx failed with error %d", ret);
+		return ret;
+	}
+
+	/* now wait for completion */
+
+	return ipc4_tx_wait_done(ipc, msg, reply_data);
+}
+
+int sof_ipc4_tx_message(struct snd_sof_ipc *ipc, u32 header, u32 extension,
+			void *msg_data, size_t msg_bytes, void *reply_data,
+			size_t reply_bytes)
+{
+	int ret;
+
+	/* Serialise IPC TX */
+	mutex_lock(&ipc->tx_mutex);
+
+	ret = sof_ipc4_tx_message_unlocked(ipc, header, extension, msg_data, msg_bytes,
+					   reply_data, reply_bytes);
+
+	mutex_unlock(&ipc->tx_mutex);
+
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc4_tx_message);
+
+void snd_sof_ipc4_msgs_rx(struct snd_sof_dev *sdev, u32 msg, u32 msg_ext)
+{
+	int err;
+
+	if (!SOF_IPC4_GLB_NOTIFY_MSG_TYPE(msg))
+		return;
+
+	switch (SOF_IPC4_GLB_NOTIFY_TYPE(msg)) {
+	case SOF_IPC4_GLB_NOTIFY_FW_READY:
+		/* check for FW boot completion */
+		if (sdev->fw_state == SOF_FW_BOOT_IN_PROGRESS) {
+			err = sof_ops(sdev)->fw_ready(sdev, msg);
+			if (err < 0)
+				sdev->fw_state = SOF_FW_BOOT_READY_FAILED;
+			else
+				sdev->fw_state = SOF_FW_BOOT_COMPLETE;
+
+			/* wake up firmware loader */
+			wake_up(&sdev->boot_wait);
+		}
+
+		break;
+	default:
+		break;
+	}
+}
+EXPORT_SYMBOL(snd_sof_ipc4_msgs_rx);
+
+int sof_ipc4_init_msg_memory(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_ipc_msg *msg;
+
+	msg = &sdev->ipc->msg;
+
+	//TODO: get max_payload_size from firmware
+	msg->max_payload_size = SOF_IPC4_MSG_MAX_SIZE;
+
+	msg->msg_data = devm_kzalloc(sdev->dev, msg->max_payload_size, GFP_KERNEL);
+	if (!msg->msg_data)
+		return -ENOMEM;
+
+	msg->reply_data = devm_kzalloc(sdev->dev, msg->max_payload_size, GFP_KERNEL);
+	if (!msg->reply_data)
+		return -ENOMEM;
+
+	/*
+	 * TODO: Currently we are not able to wake DSP up once it is suspeded.
+	 * Calling pm_runtime_get_sync() here to prevent it from entering
+	 * suspend. This should be removed once power management is implemented
+	 */
+	pm_runtime_get_sync(sdev->dev);
```

Comment:
You will need a balancing `pm_runtime_put()` on the `snd_sof_ipc_free()` path...

Reply:
> You will need a balancing `pm_runtime_put()` on the `snd_sof_ipc_free()` path...

Indeed, thanks for the reminding. I added pm_runtime_put() in snd_sof_fw_unload() since sof_ipc4_init_msg_memory() will be called in firmware load.

Reply:
Why not disable runtime PM using the debug options with IPC4 instead of adding this?

Reply:
agree with the others, we can't take code like that, or you put it is a separate patch that gets reverted with a fix later on.

Reply:
@ranj063 @plbossart 

I prefer a separate patch. Should we add "NOT FOR UPSTREAM" in the commit title?

Reply:
fine with me.